src/bolos/os_eip2333.c

```diff
@@ -0,0 +1,208 @@
+  memcpy(used_salt, HKDF_MOD_R_SALT, sizeof(HKDF_MOD_R_SALT));
+  salt_len = sizeof(HKDF_MOD_R_SALT);
+  memset(sk, 0, KEY_LENGTH);
+
+  while (cx_math_is_zero(sk, KEY_LENGTH)) {
+    sys_cx_hash_sha256(used_salt, salt_len, salt, CX_SHA256_SIZE);
+    cx_hkdf_extract(CX_SHA256, ikm, ikm_len, salt, CX_SHA256_SIZE, prk);
+    cx_hkdf_expand(CX_SHA256, prk, CX_SHA256_SIZE, key_info, key_info_len, okm, CX_HKDF_MOD_R_CEIL);
+    cx_math_mod(okm, CX_HKDF_MOD_R_CEIL, domain->n, CX_HKDF_MOD_R_CEIL);
+    memmove(sk, okm + (CX_HKDF_MOD_R_CEIL - KEY_LENGTH), KEY_LENGTH);
+    salt_len = sizeof(salt);
+    memcpy(used_salt, salt, salt_len);
+  }
+}
+
+// Derive the secret key of the master node
+// Input: seed = the source of entropy for the entire tree (at least 256 bits)
+//        seed_len = the seed length
+// Output: sk = the secret key of the master node
+
+static void cx_derive_master_sk(const unsigned char *seed, unsigned int seed_len, unsigned char *sk) {
+  unsigned char      key_info[2] = {0x00, 0x30};
+  unsigned char      seed_intern[CX_EIP2333_SEED_LENGTH] = {0};
+
+  memmove(seed_intern, seed, seed_len);
+  seed_intern[seed_len] = 0;
+  cx_hkdf_mod_r(seed_intern, seed_len+1, key_info, 2, sk);
+
+}
+
+// Derive the secret key of the child node
+// Input: parent_sk = the secret key of the parent node
+//        path = path to the index of the desired child node
+//        path_len = path length
+// Output: child_sk = the secret key of the child node
+
+static int cx_derive_child_sk(const unsigned char *parent_sk, const unsigned int *path, unsigned int path_len, unsigned char *child_sk) {
+  unsigned char  lamport_pk[LAMPORT_KEY_LENGTH];
+  unsigned char  info[2];
+  unsigned int   i;
+
+  if ((path == NULL) || (path_len < 1)) {
+    warnx("eip2333 key generation: invalid path");
+    return -1;
+  }
+  for (i = 0; i < path_len; i++) {
+    if (path[i] > 0xffffffff) {
+      warnx("eip2333 key generation: invalid path");
+      return -1;
+    }
+    cx_parent_sk_to_lamport_pk((i == 0 ? parent_sk : child_sk), path[i], lamport_pk);
+    lamport_pk[KEY_LENGTH] = 0;
+    info[0]        = 0x00;
+    info[1]        = 0x30;
+    cx_hkdf_mod_r(lamport_pk, 33, info, 2, child_sk);
+  }
+
+  return 0;
+}
+
+unsigned long sys_os_perso_derive_eip2333(cx_curve_t curve,
+                                          const unsigned int* path,
+                                          unsigned int pathLength,
+                                          unsigned char *privateKey) {
+
+  unsigned char sk[KEY_LENGTH];
+  ssize_t       seed_size;
+  uint8_t       seed[MAX_SEED_SIZE];
+  int           ret;
+  
+  if (curve != CX_CURVE_BLS12_381_G1) {
+    THROW(EXCEPTION);
+  }
+
+  seed_size = get_seed_from_env("SPECULOS_SEED", seed, sizeof(seed));
+  if (seed_size < 0) {
+    THROW(EXCEPTION);
+  }
+  cx_derive_master_sk(seed, seed_size, sk);
+  if (privateKey != NULL) {
+    ret = cx_derive_child_sk(sk, path, pathLength, privateKey);
+  }
+
+  if (ret < 0) {
```

Comment:
`ret` is not initialized if `privateKey` is `NULL`. I don't understand why there is no warning from the compiler.